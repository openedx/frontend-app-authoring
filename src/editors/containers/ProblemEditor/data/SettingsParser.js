import _ from 'lodash';

import { ShowAnswerTypes, RandomizationTypesKeys } from '../../../data/constants/problem';

export const popuplateItem = (parentObject, itemName, statekey, metadata, defaultValue = null, allowNull = false) => {
  let parent = parentObject;
  const item = _.get(metadata, itemName, null);

  // if item is null, undefined, or empty string, use defaultValue
  const finalValue = (!_.isNil(item) && item !== '') ? item : defaultValue;

  if (allowNull || (!_.isNil(finalValue) && finalValue !== defaultValue)) {
    parent = { ...parentObject, [statekey]: finalValue };
  }
  return parent;
};

export const parseScoringSettings = (metadata, defaultSettings) => {
  let scoring = {};

  const attempts = popuplateItem({}, 'max_attempts', 'number', metadata);
  const initialAttempts = _.get(attempts, 'number', null);
  const defaultAttempts = _.get(defaultSettings, 'max_attempts', null);
  attempts.unlimited = false;

  // isFinite checks if value is a finite primitive number.
  if (!_.isFinite(initialAttempts) || initialAttempts === defaultAttempts) {
    // set number to null in any case as lms will pick default value if it exists.
    attempts.number = null;
  }

  // if both block number and default number are null set unlimited to true.
  if (_.isNil(initialAttempts) && _.isNil(defaultAttempts)) {
    attempts.unlimited = true;
  }

  if (attempts.number < 0) {
    attempts.number = 0;
  }

  scoring = { ...scoring, attempts };

  scoring = popuplateItem(scoring, 'weight', 'weight', metadata);

  return scoring;
};

export const parseShowAnswer = (metadata) => {
  let showAnswer = {};

  const showAnswerType = _.get(metadata, 'showanswer', {});
  if (!_.isNil(showAnswerType) && showAnswerType in ShowAnswerTypes) {
    showAnswer = { ...showAnswer, on: showAnswerType };
  }

  showAnswer = popuplateItem(showAnswer, 'attempts_before_showanswer_button', 'afterAttempts', metadata);

  return showAnswer;
};

export const parseSettings = (metadata, defaultSettings) => {
  let settings = {};

  if (_.isNil(metadata) || _.isEmpty(metadata)) {
    return settings;
  }

  const scoring = parseScoringSettings(metadata, defaultSettings);
  if (!_.isEmpty(scoring)) {
    settings = { ...settings, scoring };
  }

  const showAnswer = parseShowAnswer(metadata);
  if (!_.isEmpty(showAnswer)) {
    settings = { ...settings, showAnswer };
  }

  const randomizationType = _.get(metadata, 'rerandomize', {});
  if (!_.isEmpty(randomizationType) && Object.values(RandomizationTypesKeys).includes(randomizationType)) {
    settings = popuplateItem(settings, 'rerandomize', 'randomization', metadata);
  }

  settings = popuplateItem(settings, 'show_reset_button', 'showResetButton', metadata);
  settings = popuplateItem(settings, 'submission_wait_seconds', 'timeBetween', metadata);
  return settings;
};
