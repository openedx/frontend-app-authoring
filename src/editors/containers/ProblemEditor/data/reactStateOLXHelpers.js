import { flatten } from 'lodash';

/**
 * flattenSubNodes - appends a nodes children to a given array 'subNodes' in a flattened format.
 * Only appends direct children unless `recursive` is set to true.
 * @param {object} node
 * @param {object[]} subNodes
 * @param {object} options - { recursive: false } Whether to search recursively for child tags.
 * @returns {void}
 */
function flattenSubNodes(node, subNodes, options = { recursive: false }) {
  const values = Array.isArray(node) ? (
    flatten(node.map(n => Object.values(n)))
  ) : Object.values(node);

  values.forEach(value => {
    if (Array.isArray(value)) {
      subNodes.push(...value);

      if (options.recursive) {
        value.forEach(nestedNode => {
          flattenSubNodes(nestedNode, subNodes, options);
        });
      }
    }
  });
}

/**
 * function nodeContainsChildTags - checks whether a node contains any subnodes with the specified tag names.
 *
 * @param {object} node - Example for node:
 *   {"div":
 *     [
 *       {"label":
 *         [
 *           {"#text":"def"}
 *         ]
 *       },
 *       {"div":
 *         [{
 *           label:
 *             [
 *               { '#text': 'def' },
 *             ],
 *         }],
 *       },
 *       {"#text":" "},
 *       {"em":
 *         [
 *           {"#text":"ghi"}
 *         ],
 *         ":@": {"@_class":"olx_description"}
 *       },
 *       {"em":
 *         [
 *           {"#text":"jkl"}
 *         ]
 *       }
 *     ]
 *   }
 * @param {string[]} tagNames
 * @param {boolean} [recursive=false] - Whether to search recursively for child tags.
 *
 * @returns {boolean} whether node contains the specified child tags
 */
export function nodeContainsChildTags(node, tagNames, options = { recursive: false }) {
  const subNodes = [];
  flattenSubNodes(node, subNodes, options);
  const res = subNodes.some(subNode => tagNames.includes(Object.keys(subNode)[0]));
  return res;
}

/**
 * @param {object} node
 * @returns {string} the tag name of the node
 */
export function tagName(node) {
  if (Array.isArray(node)) {
    throw new TypeError('function tagName does not accept arrays as input');
  }
  return Object.keys(node).find(key => key.match(/^[a-zA-Z].*/));
}

/**
 * findNodesAndRemoveTheirParentNodes - given 'arrayOfNodes', an array that results from the
 * XMLParser, this looks for 'nodesToFind': specific child tags (e.g. 'label'), and removes
 * their parent tags (e.g. 'div') but keeps all their content one level higher.
 * The 'parentsToRemove' array should contain a list of parent tags that will be removed.
 *
 * For example, if you have a div containing a p that contains a label and an em, and you specify
 * 'label' as the nodeToFind and 'p' as the parentToRemove, then you will get a div containing a label and an em.
 * @param {object} { arrayOfNodes, nodesToFind, parentsToRemove }
 *
 * @returns {object} - an array of nodes
 */
export function findNodesAndRemoveTheirParentNodes({
  arrayOfNodes,
  nodesToFind,
  parentsToRemove,
}) {
  const result = [];
  arrayOfNodes.forEach((node) => {
    const containsRelevantSubnodes = nodeContainsChildTags(node, nodesToFind);

    if (!containsRelevantSubnodes || !parentsToRemove.includes(tagName(node))) {
      result.push(node);
    } else {
      const subNodes = Object.values(node)[0];
      result.push(...subNodes);
    }
  });
  return result;
}
