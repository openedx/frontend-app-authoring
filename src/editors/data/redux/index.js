import { combineReducers } from 'redux';

import { StrictDict } from '../../utils';

import * as app from './app';
import * as requests from './requests';
import * as video from './video';
import * as problem from './problem';
import * as game from './game';

export { default as thunkActions } from './thunkActions';

const modules = {
  app,
  requests,
  video,
  problem,
  game,
};

const moduleProps = (propName) => Object.keys(modules).reduce(
  (obj, moduleKey) => ({ ...obj, [moduleKey]: modules[moduleKey][propName] }),
  /** @type {Record<string, any>} */({}),
);

const rootReducer = combineReducers(moduleProps('reducer'));

const actions = StrictDict(moduleProps('actions'));

const selectors = StrictDict(moduleProps('selectors'));

export { actions, selectors };

export default rootReducer;
