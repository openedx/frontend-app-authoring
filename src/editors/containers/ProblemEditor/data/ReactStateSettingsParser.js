import { XMLParser } from 'fast-xml-parser';
import _ from 'lodash';

import {
  ShowAnswerTypesKeys,
} from '../../../data/constants/problem';
import { popuplateItem } from './SettingsParser';

const SETTING_KEYS = [
  'max_attempts',
  'weight',
  'showanswer',
  'show_reset_button',
  'rerandomize',
];

class ReactStateSettingsParser {
  constructor(problemState) {
    this.problem = problemState.problem;
    this.rawOLX = problemState.rawOLX;
  }

  getSettings() {
    let settings = {};
    const { defaultSettings } = this.problem;
    const stateSettings = this.problem.settings;

    const numberOfAttemptsChoice = [
      ShowAnswerTypesKeys.AFTER_SOME_NUMBER_OF_ATTEMPTS,
      ShowAnswerTypesKeys.AFTER_ALL_ATTEMPTS,
      ShowAnswerTypesKeys.AFTER_ALL_ATTEMPTS_OR_CORRECT,
    ];

    settings = popuplateItem(settings, 'number', 'max_attempts', stateSettings.scoring.attempts, defaultSettings?.maxAttempts, true);
    settings = popuplateItem(settings, 'weight', 'weight', stateSettings.scoring);
    settings = popuplateItem(settings, 'on', 'showanswer', stateSettings.showAnswer, defaultSettings?.showanswer, true);
    if (_.includes(numberOfAttemptsChoice, stateSettings.showAnswer.on)) {
      settings = popuplateItem(settings, 'afterAttempts', 'attempts_before_showanswer_button', stateSettings.showAnswer);
    }
    settings = popuplateItem(settings, 'showResetButton', 'show_reset_button', stateSettings, defaultSettings?.showResetButton, true);
    settings = popuplateItem(settings, 'timeBetween', 'submission_wait_seconds', stateSettings);
    settings = popuplateItem(settings, 'randomization', 'rerandomize', stateSettings, defaultSettings?.rerandomize, true);

    return settings;
  }

  parseRawOlxSettings() {
    const rawOlxSettings = this.getSettings();
    const parserOptions = {
      ignoreAttributes: false,
      alwaysCreateTextNode: true,
      numberParseOptions: {
        leadingZeros: false,
        hex: false,
      },
    };
    const parser = new XMLParser(parserOptions);
    const olx = parser.parse(this.rawOLX);
    const settingAttributes = Object.keys(olx.problem).filter(tag => tag.startsWith('@_'));
    settingAttributes.forEach(attribute => {
      const attributeKey = attribute.substring(2);
      if (SETTING_KEYS.includes(attributeKey)) {
        if (attributeKey === 'max_attempts' || attributeKey === 'weight') {
          rawOlxSettings[attributeKey] = parseInt(olx.problem[attribute], 10);
        } else {
          rawOlxSettings[attributeKey] = olx.problem[attribute];
        }
      }
    });
    return rawOlxSettings;
  }
}

export default ReactStateSettingsParser;
