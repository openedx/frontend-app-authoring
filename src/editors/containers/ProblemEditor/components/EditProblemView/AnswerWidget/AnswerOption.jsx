import React, { memo } from 'react';
import { connect, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import {
  Collapsible,
  Icon,
  IconButton,
  Form,
} from '@openedx/paragon';
import { FeedbackOutline, DeleteOutline } from '@openedx/paragon/icons';
import { FormattedMessage, injectIntl, intlShape } from '@edx/frontend-platform/i18n';
import messages from './messages';
import { selectors } from '../../../../../data/redux';
import { answerOptionProps } from '../../../../../data/services/cms/types';
import Checker from './components/Checker';
import { FeedbackBox } from './components/Feedback';
import * as hooks from './hooks';
import { ProblemTypeKeys } from '../../../../../data/constants/problem';
import ExpandableTextArea from '../../../../../sharedComponents/ExpandableTextArea';

const AnswerOption = ({
  answer,
  hasSingleAnswer,
  // injected
  intl,
  // redux
  problemType,
  images,
  isLibrary,
  learningContextId,
}) => {
  const dispatch = useDispatch();
  const removeAnswer = hooks.removeAnswer({ answer, dispatch });
  const setAnswer = hooks.setAnswer({ answer, hasSingleAnswer, dispatch });
  const setAnswerTitle = hooks.setAnswerTitle({
    answer,
    hasSingleAnswer,
    dispatch,
    problemType,
  });
  const setSelectedFeedback = hooks.setSelectedFeedback({ answer, hasSingleAnswer, dispatch });
  const setUnselectedFeedback = hooks.setUnselectedFeedback({ answer, hasSingleAnswer, dispatch });
  const { isFeedbackVisible, toggleFeedback } = hooks.useFeedback(answer);

  const getInputArea = () => {
    if ([ProblemTypeKeys.SINGLESELECT, ProblemTypeKeys.MULTISELECT].includes(problemType)) {
      return (
        <ExpandableTextArea
          value={answer.title}
          setContent={setAnswerTitle}
          placeholder={intl.formatMessage(messages.answerTextboxPlaceholder)}
          id={`answer-${answer.id}`}
          {...{
            images,
            isLibrary,
            learningContextId,
          }}
        />
      );
    }
    if (problemType !== ProblemTypeKeys.NUMERIC || !answer.isAnswerRange) {
      return (
        <Form.Control
          as="textarea"
          className="answer-option-textarea text-gray-500 small"
          autoResize
          rows={1}
          value={answer.title}
          onChange={setAnswerTitle}
          placeholder={intl.formatMessage(messages.answerTextboxPlaceholder)}
        />
      );
    }
    // Return Answer Range View
    return (
      <div>
        <Form.Control
          as="textarea"
          className="answer-option-textarea text-gray-500 small"
          autoResize
          rows={1}
          value={answer.title}
          onChange={setAnswerTitle}
          placeholder={intl.formatMessage(messages.answerRangeTextboxPlaceholder)}
        />
        <div className="pgn__form-switch-helper-text">
          <FormattedMessage {...messages.answerRangeHelperText} />
        </div>
      </div>

    );
  };

  return (
    <Collapsible.Advanced
      open={isFeedbackVisible}
      onToggle={toggleFeedback}
      className="answer-option d-flex flex-row justify-content-between flex-nowrap pb-2 pt-2"
    >
      <div className="mr-1 d-flex">
        <Checker
          hasSingleAnswer={hasSingleAnswer}
          answer={answer}
          setAnswer={setAnswer}
          disabled={problemType === ProblemTypeKeys.NUMERIC}
        />
      </div>
      <div className="ml-1 flex-grow-1">
        {getInputArea()}
        <Collapsible.Body>
          <FeedbackBox
            problemType={problemType}
            answer={answer}
            setSelectedFeedback={setSelectedFeedback}
            setUnselectedFeedback={setUnselectedFeedback}
            intl={intl}
            {...{
              images,
              isLibrary,
              learningContextId,
            }}
          />
        </Collapsible.Body>
      </div>
      <div className="d-flex flex-row flex-nowrap">
        <Collapsible.Trigger aria-label="Toggle feedback" className="btn-icon btn-icon-primary btn-icon-md align-items-center">
          <Icon
            src={FeedbackOutline}
            alt={intl.formatMessage(messages.feedbackToggleIconAltText)}
          />
        </Collapsible.Trigger>
        <IconButton
          src={DeleteOutline}
          iconAs={Icon}
          alt={intl.formatMessage(messages.answerDeleteIconAltText)}
          onClick={removeAnswer}
          variant="primary"
        />
      </div>
    </Collapsible.Advanced>
  );
};

AnswerOption.propTypes = {
  answer: answerOptionProps.isRequired,
  hasSingleAnswer: PropTypes.bool.isRequired,
  // injected
  intl: intlShape.isRequired,
  // redux
  problemType: PropTypes.string.isRequired,
  images: PropTypes.shape({}).isRequired,
  learningContextId: PropTypes.string.isRequired,
  isLibrary: PropTypes.bool.isRequired,
};

export const mapStateToProps = (state) => ({
  problemType: selectors.problem.problemType(state),
  images: selectors.app.images(state),
  isLibrary: selectors.app.isLibrary(state),
  learningContextId: selectors.app.learningContextId(state),
});

export const mapDispatchToProps = {};
export const AnswerOptionInternal = AnswerOption; // For testing only
export default injectIntl(connect(mapStateToProps, mapDispatchToProps)(memo(AnswerOption)));
