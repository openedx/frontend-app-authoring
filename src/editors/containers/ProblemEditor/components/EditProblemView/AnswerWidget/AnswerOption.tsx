import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  Collapsible,
  Icon,
  IconButton,
  Form,
} from '@openedx/paragon';
import { FeedbackOutline, DeleteOutline } from '@openedx/paragon/icons';
import { FormattedMessage, useIntl } from '@edx/frontend-platform/i18n';
import messages from './messages';
import { selectors } from '../../../../../data/redux';
import Checker from './components/Checker';
import { FeedbackBox } from './components/Feedback';
import * as hooks from './hooks';
import { AnswerData, ProblemTypeKeys } from '../../../../../data/constants/problem';
import ExpandableTextArea from '../../../../../sharedComponents/ExpandableTextArea';
import { useEditorContext } from '../../../../../EditorContext';

interface Props {
  answer: AnswerData;
  hasSingleAnswer: boolean;
}

export const AnswerOption = ({
  answer,
  hasSingleAnswer,
}: Props) => {
  const intl = useIntl();
  const dispatch = useDispatch();
  const { learningContextId } = useEditorContext();
  const problemType = useSelector(selectors.problem.problemType);
  const images = useSelector(selectors.app.images);
  const isLibrary = useSelector(selectors.app.isLibrary);
  const removeAnswer = hooks.removeAnswer({ answer, dispatch });
  const setAnswer = hooks.setAnswer({ answer, hasSingleAnswer, dispatch });
  const setAnswerTitle = hooks.setAnswerTitle({
    answer,
    hasSingleAnswer,
    dispatch,
    problemType,
  });
  const setSelectedFeedback = hooks.setSelectedFeedback({ answer, hasSingleAnswer, dispatch });
  const setUnselectedFeedback = hooks.setUnselectedFeedback({ answer, hasSingleAnswer, dispatch });
  const { isFeedbackVisible, toggleFeedback } = hooks.useFeedback(answer);

  const getInputArea = () => {
    if ([ProblemTypeKeys.SINGLESELECT, ProblemTypeKeys.MULTISELECT].includes(problemType as any)) {
      return (
        <ExpandableTextArea
          value={answer.title}
          setContent={setAnswerTitle}
          placeholder={intl.formatMessage(messages.answerTextboxPlaceholder)}
          id={`answer-${answer.id}`}
          {...{
            images,
            isLibrary,
            learningContextId,
          }}
        />
      );
    }
    if (problemType !== ProblemTypeKeys.NUMERIC || !answer.isAnswerRange) {
      return (
        <Form.Control
          as="textarea"
          className="answer-option-textarea text-gray-500 small"
          autoResize
          rows={1}
          value={answer.title}
          onChange={setAnswerTitle}
          placeholder={intl.formatMessage(messages.answerTextboxPlaceholder)}
        />
      );
    }
    // Return Answer Range View
    return (
      <div>
        <Form.Control
          as="textarea"
          className="answer-option-textarea text-gray-500 small"
          autoResize
          rows={1}
          value={answer.title}
          onChange={setAnswerTitle}
          placeholder={intl.formatMessage(messages.answerRangeTextboxPlaceholder)}
        />
        <div className="pgn__form-switch-helper-text">
          <FormattedMessage {...messages.answerRangeHelperText} />
        </div>
      </div>

    );
  };

  return (
    <Collapsible.Advanced
      open={isFeedbackVisible}
      onToggle={toggleFeedback}
      className="answer-option d-flex flex-row justify-content-between flex-nowrap pb-2 pt-2"
    >
      <div className="mr-1 d-flex">
        <Checker
          hasSingleAnswer={hasSingleAnswer}
          answer={answer}
          setAnswer={setAnswer}
          disabled={problemType === ProblemTypeKeys.NUMERIC}
        />
      </div>
      <div className="ml-1 flex-grow-1">
        {getInputArea()}
        <Collapsible.Body>
          <FeedbackBox
            problemType={problemType}
            answer={answer}
            setSelectedFeedback={setSelectedFeedback}
            setUnselectedFeedback={setUnselectedFeedback}
            {...{
              images,
              isLibrary,
              learningContextId,
            }}
          />
        </Collapsible.Body>
      </div>
      <div className="d-flex flex-row flex-nowrap">
        <Collapsible.Trigger aria-label="Toggle feedback" className="btn-icon btn-icon-primary btn-icon-md align-items-center">
          <Icon
            src={FeedbackOutline}
            screenReaderText={intl.formatMessage(messages.feedbackToggleIconAltText)}
          />
        </Collapsible.Trigger>
        <IconButton
          src={DeleteOutline}
          iconAs={Icon}
          alt={intl.formatMessage(messages.answerDeleteIconAltText)}
          onClick={removeAnswer}
          variant="primary"
        />
      </div>
    </Collapsible.Advanced>
  );
};
