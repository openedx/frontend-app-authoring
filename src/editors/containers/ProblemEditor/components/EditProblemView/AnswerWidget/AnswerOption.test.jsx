import 'CourseAuthoring/editors/setupEditorTest';
import React from 'react';
import { shallow } from '@edx/react-unit-test-utils';
import { formatMessage } from '../../../../../testUtils';
import { selectors } from '../../../../../data/redux';
import { AnswerOptionInternal as AnswerOption, mapStateToProps } from './AnswerOption';

jest.mock('../../../../../data/redux', () => ({
  __esModule: true,
  default: jest.fn(),
  selectors: {
    problem: {
      problemType: jest.fn(state => ({ problemType: state })),
    },
    app: {
      images: jest.fn(state => ({ images: state })),
      isLibrary: jest.fn(state => ({ isLibrary: state })),
      learningContextId: jest.fn(state => ({ learningContextId: state })),
    },
  },
  thunkActions: {
    video: {
      importTranscripts: jest.fn(),
    },
  },
}));

describe('AnswerOption', () => {
  const answerWithOnlyFeedback = {
    id: 'A',
    title: 'Answer 1',
    correct: true,
    selectedFeedback: 'some feedback',
  };
  const answerWithSelectedUnselectedFeedback = {
    id: 'A',
    title: 'Answer 1',
    correct: true,
    selectedFeedback: 'selected feedback',
    unselectedFeedback: 'unselected feedback',
  };
  const answerRange = {
    id: 'A',
    title: 'Answer 1',
    correct: true,
    selectedFeedback: 'selected feedback',
    unselectedFeedback: 'unselected feedback',
    isAnswerRange: true,
  };

  const props = {
    hasSingleAnswer: false,
    answer: answerWithOnlyFeedback,
    // inject
    intl: { formatMessage },
    // redux
    problemType: 'multiplechoiceresponse',
    images: {},
    isLibrary: false,
    learningContextId: 'course+org+run',
  };
  describe('render', () => {
    test('snapshot: renders correct option with feedback', () => {
      expect(shallow(<AnswerOption {...props} />).snapshot).toMatchSnapshot();
    });
    test('snapshot: renders correct option with selected unselected feedback', () => {
      expect(shallow(<AnswerOption {...props} problemType="choiceresponse" answer={answerWithSelectedUnselectedFeedback} />).snapshot).toMatchSnapshot();
    });
    test('snapshot: renders correct option with numeric input problem', () => {
      expect(shallow(<AnswerOption {...props} problemType="numericalresponse" />).snapshot).toMatchSnapshot();
    });
    test('snapshot: renders correct option with numeric input problem and answer range', () => {
      expect(shallow(<AnswerOption {...props} problemType="numericalresponse" answer={answerRange} />).snapshot).toMatchSnapshot();
    });
  });

  describe('mapStateToProps', () => {
    const testState = { A: 'pple', B: 'anana', C: 'ucumber' };
    test('problemType from problem.problemType', () => {
      expect(
        mapStateToProps(testState).problemType,
      ).toEqual(selectors.problem.problemType(testState));
    });
    test('images from app.images', () => {
      expect(
        mapStateToProps(testState).images,
      ).toEqual(selectors.app.images(testState));
    });
    test('learningContextId from app.learningContextId', () => {
      expect(
        mapStateToProps(testState).learningContextId,
      ).toEqual(selectors.app.learningContextId(testState));
    });
    test('isLibrary from app.isLibrary', () => {
      expect(
        mapStateToProps(testState).isLibrary,
      ).toEqual(selectors.app.isLibrary(testState));
    });
  });
});
