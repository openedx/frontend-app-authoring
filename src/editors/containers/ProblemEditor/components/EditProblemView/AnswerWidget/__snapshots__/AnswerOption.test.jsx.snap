// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnswerOption render snapshot: renders correct option with feedback 1`] = `
<Advanced
  className="answer-option d-flex flex-row justify-content-between flex-nowrap pb-2 pt-2"
  onToggle={[Function]}
  open={false}
>
  <div
    className="mr-1 d-flex"
  >
    <Checker
      answer={
        {
          "correct": true,
          "id": "A",
          "selectedFeedback": "some feedback",
          "title": "Answer 1",
        }
      }
      disabled={false}
      hasSingleAnswer={false}
      setAnswer={[Function]}
    />
  </div>
  <div
    className="ml-1 flex-grow-1"
  >
    <ExpandableTextArea
      error={false}
      errorMessage={null}
      id="answer-A"
      images={{}}
      isLibrary={false}
      learningContextId="course+org+run"
      placeholder="Enter an answer"
      setContent={[Function]}
      value="Answer 1"
    />
    <Body>
      <injectIntl(ShimmedIntlComponent)
        answer={
          {
            "correct": true,
            "id": "A",
            "selectedFeedback": "some feedback",
            "title": "Answer 1",
          }
        }
        images={{}}
        intl={
          {
            "formatMessage": [Function],
          }
        }
        isLibrary={false}
        learningContextId="course+org+run"
        problemType="multiplechoiceresponse"
        setSelectedFeedback={[Function]}
        setUnselectedFeedback={[Function]}
      />
    </Body>
  </div>
  <div
    className="d-flex flex-row flex-nowrap"
  >
    <Trigger
      aria-label="Toggle feedback"
      className="btn-icon btn-icon-primary btn-icon-md align-items-center"
    >
      <Icon
        alt="Toggle feedback"
      />
    </Trigger>
    <IconButton
      alt="Delete answer"
      iconAs="Icon"
      onClick={[Function]}
      variant="primary"
    />
  </div>
</Advanced>
`;

exports[`AnswerOption render snapshot: renders correct option with numeric input problem 1`] = `
<Advanced
  className="answer-option d-flex flex-row justify-content-between flex-nowrap pb-2 pt-2"
  onToggle={[Function]}
  open={false}
>
  <div
    className="mr-1 d-flex"
  >
    <Checker
      answer={
        {
          "correct": true,
          "id": "A",
          "selectedFeedback": "some feedback",
          "title": "Answer 1",
        }
      }
      disabled={true}
      hasSingleAnswer={false}
      setAnswer={[Function]}
    />
  </div>
  <div
    className="ml-1 flex-grow-1"
  >
    <Form.Control
      as="textarea"
      autoResize={true}
      className="answer-option-textarea text-gray-500 small"
      onChange={[Function]}
      placeholder="Enter an answer"
      rows={1}
      value="Answer 1"
    />
    <Body>
      <injectIntl(ShimmedIntlComponent)
        answer={
          {
            "correct": true,
            "id": "A",
            "selectedFeedback": "some feedback",
            "title": "Answer 1",
          }
        }
        images={{}}
        intl={
          {
            "formatMessage": [Function],
          }
        }
        isLibrary={false}
        learningContextId="course+org+run"
        problemType="numericalresponse"
        setSelectedFeedback={[Function]}
        setUnselectedFeedback={[Function]}
      />
    </Body>
  </div>
  <div
    className="d-flex flex-row flex-nowrap"
  >
    <Trigger
      aria-label="Toggle feedback"
      className="btn-icon btn-icon-primary btn-icon-md align-items-center"
    >
      <Icon
        alt="Toggle feedback"
      />
    </Trigger>
    <IconButton
      alt="Delete answer"
      iconAs="Icon"
      onClick={[Function]}
      variant="primary"
    />
  </div>
</Advanced>
`;

exports[`AnswerOption render snapshot: renders correct option with numeric input problem and answer range 1`] = `
<Advanced
  className="answer-option d-flex flex-row justify-content-between flex-nowrap pb-2 pt-2"
  onToggle={[Function]}
  open={false}
>
  <div
    className="mr-1 d-flex"
  >
    <Checker
      answer={
        {
          "correct": true,
          "id": "A",
          "isAnswerRange": true,
          "selectedFeedback": "selected feedback",
          "title": "Answer 1",
          "unselectedFeedback": "unselected feedback",
        }
      }
      disabled={true}
      hasSingleAnswer={false}
      setAnswer={[Function]}
    />
  </div>
  <div
    className="ml-1 flex-grow-1"
  >
    <div>
      <Form.Control
        as="textarea"
        autoResize={true}
        className="answer-option-textarea text-gray-500 small"
        onChange={[Function]}
        placeholder="Enter an answer range"
        rows={1}
        value="Answer 1"
      />
      <div
        className="pgn__form-switch-helper-text"
      >
        <FormattedMessage
          defaultMessage="Enter min and max values separated by a comma. Use a bracket to include the number next to it in the range, or a parenthesis to exclude the number. For example, to identify the correct answers as 5, 6, or 7, but not 8, specify [5,8)."
          description="Helper text describing usage of answer ranges"
          id="authoring.answerwidget.answer.answerRangeHelperText"
        />
      </div>
    </div>
    <Body>
      <injectIntl(ShimmedIntlComponent)
        answer={
          {
            "correct": true,
            "id": "A",
            "isAnswerRange": true,
            "selectedFeedback": "selected feedback",
            "title": "Answer 1",
            "unselectedFeedback": "unselected feedback",
          }
        }
        images={{}}
        intl={
          {
            "formatMessage": [Function],
          }
        }
        isLibrary={false}
        learningContextId="course+org+run"
        problemType="numericalresponse"
        setSelectedFeedback={[Function]}
        setUnselectedFeedback={[Function]}
      />
    </Body>
  </div>
  <div
    className="d-flex flex-row flex-nowrap"
  >
    <Trigger
      aria-label="Toggle feedback"
      className="btn-icon btn-icon-primary btn-icon-md align-items-center"
    >
      <Icon
        alt="Toggle feedback"
      />
    </Trigger>
    <IconButton
      alt="Delete answer"
      iconAs="Icon"
      onClick={[Function]}
      variant="primary"
    />
  </div>
</Advanced>
`;

exports[`AnswerOption render snapshot: renders correct option with selected unselected feedback 1`] = `
<Advanced
  className="answer-option d-flex flex-row justify-content-between flex-nowrap pb-2 pt-2"
  onToggle={[Function]}
  open={false}
>
  <div
    className="mr-1 d-flex"
  >
    <Checker
      answer={
        {
          "correct": true,
          "id": "A",
          "selectedFeedback": "selected feedback",
          "title": "Answer 1",
          "unselectedFeedback": "unselected feedback",
        }
      }
      disabled={false}
      hasSingleAnswer={false}
      setAnswer={[Function]}
    />
  </div>
  <div
    className="ml-1 flex-grow-1"
  >
    <ExpandableTextArea
      error={false}
      errorMessage={null}
      id="answer-A"
      images={{}}
      isLibrary={false}
      learningContextId="course+org+run"
      placeholder="Enter an answer"
      setContent={[Function]}
      value="Answer 1"
    />
    <Body>
      <injectIntl(ShimmedIntlComponent)
        answer={
          {
            "correct": true,
            "id": "A",
            "selectedFeedback": "selected feedback",
            "title": "Answer 1",
            "unselectedFeedback": "unselected feedback",
          }
        }
        images={{}}
        intl={
          {
            "formatMessage": [Function],
          }
        }
        isLibrary={false}
        learningContextId="course+org+run"
        problemType="choiceresponse"
        setSelectedFeedback={[Function]}
        setUnselectedFeedback={[Function]}
      />
    </Body>
  </div>
  <div
    className="d-flex flex-row flex-nowrap"
  >
    <Trigger
      aria-label="Toggle feedback"
      className="btn-icon btn-icon-primary btn-icon-md align-items-center"
    >
      <Icon
        alt="Toggle feedback"
      />
    </Trigger>
    <IconButton
      alt="Delete answer"
      iconAs="Icon"
      onClick={[Function]}
      variant="primary"
    />
  </div>
</Advanced>
`;
